Add Apache Phoenix connector

vincentpoon opened the pull request:
https://phoenix.apache.org/

Ref #76 
Ref https://github.com/prestodb/presto/pull/12111

vincentpoon commented:
This is continuing from #76 since the GitHub thread there got too large

electrum commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,276 @@
+            <artifactId>joda-time</artifactId>
+        </dependency>
+
+        <dependency>
```
This shouldn't be needed as it's already brought in by Airlift `log-manager` which is used by `bootstrap`

electrum commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,276 @@
+            <artifactId>jackson-annotations</artifactId>
+            <scope>provided</scope>
+        </dependency>
+
```
Add a comment to delineate this section
```xml
<!-- for testing -->

electrum commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,276 @@
+        <dependency>
+            <groupId>org.apache.hbase</groupId>
+            <artifactId>hbase-common</artifactId>
+            <version>1.4.0</version>
```
Add a property `dep.hbase.version` since (I'm assuming) all of these need to have the same version

electrum commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,276 @@
+                    </ignoredResourcePatterns>
+                </configuration>
+            </plugin>
+
```
Nit: remove extra blank line

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplit.java:
```diff
@@ -0,0 +1,73 @@
+        return phoenixInputSplit.getPhoenixInputSplit();
+    }
+
+    @JsonProperty("constraint")
```
No need for the name here since it is the same as the default one

electrum commented on presto-phoenix/pom.xml:
```diff
@@ -0,0 +1,276 @@
+                <groupId>org.basepom.maven</groupId>
+                <artifactId>duplicate-finder-maven-plugin</artifactId>
+                <configuration>
+                    <ignoredDependencies>
```
This is bad, since the classes are actually duplicated in the Phoenix client, they are different versions, and which one you get at runtime is effectively arbitrary. 

Ideally, we could fix the shaded Phoenix client dependency not to include these. But for now, I think the best approach is to exclude the dependencies everywhere they appear, so that we'll use the classes from the shaded JAR (and hope the versions of those classes are ok). This is ugly, but better than duplicate classes on the class path.

```diff
diff --git a/presto-phoenix/pom.xml b/presto-phoenix/pom.xml
index 5af401c086..04874679ec 100644
--- a/presto-phoenix/pom.xml
+++ b/presto-phoenix/pom.xml
@@ -16,6 +16,30 @@
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
     </properties>
 
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>io.airlift</groupId>
+                <artifactId>log-manager</artifactId>
+                <scope>runtime</scope>
+                <exclusions>
+                    <exclusion>
+                        <groupId>org.slf4j</groupId>
+                        <artifactId>jcl-over-slf4j</artifactId>
+                    </exclusion>
+                    <exclusion>
+                        <groupId>org.slf4j</groupId>
+                        <artifactId>log4j-over-slf4j</artifactId>
+                    </exclusion>
+                    <exclusion>
+                        <groupId>org.slf4j</groupId>
+                        <artifactId>slf4j-api</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+
     <dependencies>
         <dependency>
             <groupId>org.apache.phoenix</groupId>
@@ -84,12 +108,6 @@
             <artifactId>joda-time</artifactId>
         </dependency>
 
-        <dependency>
-            <groupId>org.slf4j</groupId>
-            <artifactId>jcl-over-slf4j</artifactId>
-            <scope>runtime</scope>
-        </dependency>
-
         <!-- Presto SPI -->
         <dependency>
             <groupId>io.prestosql</groupId>
@@ -123,6 +141,10 @@
                     <groupId>org.apache.yetus</groupId>
                     <artifactId>audience-annotations</artifactId>
                 </exclusion>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-api</artifactId>
+                </exclusion>
             </exclusions>
             <scope>test</scope>
         </dependency>
@@ -143,6 +165,12 @@
             <groupId>io.prestosql</groupId>
             <artifactId>presto-main</artifactId>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-api</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
 
         <dependency>
@@ -161,6 +189,12 @@
             <groupId>io.prestosql</groupId>
             <artifactId>presto-tests</artifactId>
             <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-api</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
 
         <dependency>
@@ -246,20 +280,6 @@
                 <groupId>org.basepom.maven</groupId>
                 <artifactId>duplicate-finder-maven-plugin</artifactId>
                 <configuration>
-                    <ignoredDependencies>
-                        <dependency>
-                            <groupId>org.slf4j</groupId>
-                            <artifactId>jcl-over-slf4j</artifactId>
-                        </dependency>
-                        <dependency>
-                            <groupId>org.slf4j</groupId>
-                            <artifactId>log4j-over-slf4j</artifactId>
-                        </dependency>
-                        <dependency>
-                            <groupId>org.slf4j</groupId>
-                            <artifactId>slf4j-api</artifactId>
-                        </dependency>
-                    </ignoredDependencies>
                     <ignoredResourcePatterns>
                         <ignoredResourcePattern>mrapp-generated-classpath</ignoredResourcePattern>
                         <ignoredResourcePattern>about.html</ignoredResourcePattern>
```

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+                        config.getConnectionProperties()));
+        requireNonNull(config, "config is null");
+        configuration = new Configuration(false);
+        ConnectionInfo connectionInfo = PhoenixEmbeddedDriver.ConnectionInfo.create(config.getConnectionUrl());
```
Nit: the `PhoenixEmbeddedDriver.` prefix is redundant since `ConnectionInfo` is already imported directly

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+                        Optional.empty(),
+                        Optional.empty(),
+                        config.getConnectionProperties()));
+        requireNonNull(config, "config is null");
```
This check is not needed. If `config` is null, the usage in the `super()` arguments above would have already thrown.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+    public PhoenixConnection getConnection(JdbcIdentity identity)
+            throws SQLException
+    {
+        Connection connection = connectionFactory.openConnection(identity);
```
This try/catch seems to be wrong. I would assume that `openConnection()` is what will throw the error (not the `unwrap` call). Since `getConnection()` can throw `SQLException`, we should probably just remove the try/catch here and let the caller deal with it.

vincentpoon replied:
There was concern in a prior review that in the (very) unlikely case that unwrap fails, the connection would never get closed.  Hence the try/catch here

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+    @Override
+    protected Collection<String> listSchemas(Connection connection)
+    {
+        try (ResultSet resultSet = connection.getMetaData().getSchemas()) {
```
I think this method can be simplified to
```java
return ImmutableSet.builder()
        .add(DEFAULT_SCHEMA)
        .addAll(super.listSchemas(connection))
        .build();
```
(this would return `JDBC_ERROR` on error but that seems fine)

vincentpoon replied:
There is another subtle difference from the parent class implementation here - we check if `schemaName` is `null`.  Phoenix can return `null` in `connection.getMetaData.getSchemas`, so without this check we'll get a NPE.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+        return Optional.of((sql, limit) -> sql + " LIMIT " + limit);
+    }
+
+    private Function<String, String> tryApplyLimit(OptionalLong limit)
```
Remove this method and make the one in `BaseJdbcClient` protected

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+            throws SQLException
+    {
+        String schemaName = resultSet.getString("TABLE_SCHEM");
+        return schemaName == null ? DEFAULT_SCHEMA : schemaName;
```
Can simplify to
```java
return firstNonNull(resultSet.getString("TABLE_SCHEM"), DEFAULT_SCHEMA);

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+        return schemaName == null ? DEFAULT_SCHEMA : schemaName;
+    }
+
+    protected static ResultSet getColumns(JdbcTableHandle tableHandle, DatabaseMetaData metadata)
```
This method is not used

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+                        .map(elementMapping -> {
+                            ArrayType prestoArrayType = new ArrayType(elementMapping.getType());
+                            String jdbcTypeName = elementTypeHandle.getJdbcTypeName()
+                                    .orElseThrow(() -> new PrestoException(PHOENIX_METADATA_ERROR, "Type name is missing: " + typeHandle));
```
The `toString()` for `JdbcTypeHandle` is not friendly for end users. We might want to improve this if we expect that this can happen in practice.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/TypeUtils.java:
```diff
@@ -0,0 +1,241 @@
+
+    public static Object[] toBoxedArray(Object jdbcArray)
+    {
+        if (!jdbcArray.getClass().isArray()) {
```
Probably better to have this be
```java
checkArgument(jdbc.getClass().isArray(), "object is not an array: %s", jdbcArray.getClass().getName());

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+        String arrayTypeName = arrayTypeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new PrestoException(JDBC_ERROR, "Type name is missing: " + arrayTypeHandle));
+        checkArgument(arrayTypeName.endsWith(" ARRAY"), "array type must end with ' ARRAY'");
+        arrayTypeName = arrayTypeName.substring(0, arrayTypeName.lastIndexOf(" ARRAY"));
```
We don't need to search since we already know it's at the end
```java
arrayTypeName = arrayTypeName.substring(0, arrayTypeName.length() - " ARRAY".length());

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+        }
+    }
+
+    private ResultSet getResultSet(InputSplit split, QueryPlan queryPlan)
```
This method can be static

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+            ConnectionQueryServices services = context.getConnection().getQueryServices();
+            services.clearTableRegionCache(tableNameBytes);
+
+            long renewScannerLeaseThreshold = context.getConnection().getQueryServices().getRenewLeaseThresholdMilliSeconds();
```
Can reuse the `services` variable

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+                ScanMetricsHolder scanMetricsHolder =
+                        ScanMetricsHolder.getInstance(
+                                context.getReadMetricsQueue(),
+                                queryPlan.getTableRef().getTable().getPhysicalName().getString(), scan,
```
Nit: the code style is to put all arguments on one line, or to wrap each argument. So put `scan` on a separate line.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+
+                TableResultIterator tableResultIterator =
+                        new TableResultIterator(
+                                context.getConnection().getMutationState(), scan,
```
Same here, one argument per line

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+                                context.getConnection().getLogLevel());
+
+                TableResultIterator tableResultIterator =
+                        new TableResultIterator(
```
I'd format this on the previous line
```java
TableResultIterator tableResultIterator = new TableResultIterator(
        arg1,
        arg2,
        ...

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,453 @@
+                // For MR, skip the region boundary check exception if we encounter a split. ref: PHOENIX-2599
+                scan.setAttribute(SKIP_REGION_BOUNDARY_CHECK, Bytes.toBytes(true));
+
+                PeekingResultIterator peekingResultIterator;
```
Move the variable declaration below to where it is initialized, or simply inline the expression (it is short)

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixOutputTableHandle.java:
```diff
@@ -0,0 +1,48 @@
+public class PhoenixOutputTableHandle
+        extends JdbcOutputTableHandle
+{
+    private final boolean hasUUIDRowkey;
```
Nit: "row key" seems to be two words per the Phoenix docs

Also, for our code style, we only capitalize the first letter of acronyms, so `HTTP` -> `Http`, `URL` -> `Url`, etc. This makes it easier to read when other words follow acronyms, or you have two acronyms in a row.

So this should probably be `hasUuidRowKey`

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixPlugin.java:
```diff
@@ -0,0 +1,35 @@
+
+    private static ClassLoader getClassLoader()
+    {
+        return firstNonNull(Thread.currentThread().getContextClassLoader(), PhoenixPlugin.class.getClassLoader());
```
Remove this method and use `PhoenixClient.class.getClassLoader()` above. That one is always correct -- we shouldn't rely on the thread context class loader, which is intended for other purposes. (we only set it for interoperability with Hadoop code)

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixSplitManager.java:
```diff
@@ -0,0 +1,158 @@
+                    Optional.empty(),
+                    Function.identity());
+
+            return new FixedSplitSource(getSplits(inputQuery).stream().map(PhoenixInputSplit.class::cast).map(split ->
```
This is a bit hard to read. Let's separate it out
```java
List<ConnectorSplit> splits = getSplits(inputQuery).stream()
        .map(PhoenixInputSplit.class::cast)
        .map(split -> new PhoenixSplit(
                ...))
        .collect(toImmutableList());
return new FixedSplitSource(splits);

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixTableProperties.java:
```diff
@@ -0,0 +1,266 @@
+/**
+ * Class contains all table properties for the Phoenix connector. Used when creating a table:
+ * <p>
+ * CREATE TABLE foo (a VARCHAR with (primary_key = true), b INT) WITH (SALT_BUCKETS=10, VERSIONS=5, COMPRESSION='lz');
```
Nit: add a `<pre>` tag around this

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixTableProperties.java:
```diff
@@ -0,0 +1,266 @@
+ */
+public final class PhoenixTableProperties
+{
+    private final List<PropertyMetadata<?>> tableProperties;
```
Nit: put the class field below the constants and add a newline to separate them

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixTableProperties.java:
```diff
@@ -0,0 +1,266 @@
+    public static final String COMPRESSION = "compression";
+    public static final String TTL = "ttl";
+
+    private final List<PropertyMetadata<?>> columnProperties;
```
Let's split this out into a `PhoenixColumnProperties` class to avoid confusion about why column properties are in the table properties class

electrum commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/PhoenixQueryRunner.java:
```diff
@@ -0,0 +1,135 @@
+                Statement statement = connection.createStatement()) {
+            statement.execute(format("CREATE SCHEMA %s", schema));
+        }
+        catch (Exception e) {
```
Let's remove the catch and make this method throw `SQLException`

electrum commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixDistributedQueries.java:
```diff
@@ -0,0 +1,118 @@
+        assertQueryFails("COMMENT ON TABLE orders IS 'hello'", "This connector does not support setting table comments");
+    }
+
+    @AfterClass(alwaysRun = true)
```
Move this method to the top of the class, after the constructor, so that the lifecycle stuff is together.

vincentpoon commented:
Is there a known issue with the builds?  I'm not able to repro the test failure on my local

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,455 @@
+    {
+        JdbcTableHandle existingTable = getTableHandle(session, tableMetadata.getTable());
+        if (existingTable != null && !ignoreExisting) {
+            throw new IllegalArgumentException("Target table already exists: " + tableMetadata.getTable());
```
Throw `PrestoException(ALREADY_EXISTS)` here, since this is a user error. Also, it's fine to unconditionally throw, regardless of `ignoreExisting`. That flag is there for transactional connectors like Hive that create tables later during the commit call. For non-transactional connectors, the engine's `CreateTableTask` will correctly handle this if the connector throws with `ALREADY_EXISTS`.

In practice, this only ever happens in a race condition, since the engine already checks if the table exists before calling this method. And actually, we still have a race condition, since this check happens before the actual create.

So, I would say just remove this check, since it's redundant with the one the engine performs.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,455 @@
+                    hasUUIDRowkey);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(PHOENIX_METADATA_ERROR, "Error creating Phoenix table", e);
```
As mentioned above, we should throw `ALREADY_EXISTS` if possible when the table already exists. Maybe Phoenix has an error code or well-known error message to compare against?

Throwing `ALREADY_EXISTS` here will allow `CREATE TABLE IF NOT EXISTS` to work correctly if someone else created the table after the engine already checked.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,455 @@
+    {
+        JdbcTableHandle handle = (JdbcTableHandle) tableHandle;
+        ConnectorTableMetadata tableMetadata = getTableMetadata(session, tableHandle, true);
+        List<ColumnMetadata> nonRowkeyCols = tableMetadata.getColumns().stream().filter(column -> !ROWKEY.equalsIgnoreCase(column.getName())).collect(Collectors.toList());
```
Nit: wrap the `filter` and `collect` calls on separate lines. Also, use `toImmutableList` with a static import

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,455 @@
+                handle.getTableName(),
+                nonRowkeyCols.stream().map(ColumnMetadata::getName).collect(Collectors.toList()),
+                nonRowkeyCols.stream().map(ColumnMetadata::getType).collect(Collectors.toList()),
+                "", // no use of temporary tables
```
How does this work? This field is used in several places in `BaseJdbcClient` which are not overridden here, and there is no special casing for empty.

If it is somehow never used, it would be cleaner to remove it from the constructor of `PhoenixOutputTableHandle` and pass `""` when calling `super()`. This is fine since Jackson will ignore properties that don't have a constructor parameter when deserializing.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,455 @@
+    @Override
+    public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, Optional<ConnectorNewTableLayout> layout)
+    {
+        JdbcOutputTableHandle handle = createTable(session, tableMetadata);
```
The table creation needs to happen in `finishCreateTable`. The reason is that `beginCreateTable` is called during query planning, so an `EXPLAIN CREATE TABLE xxx AS ...` or similar would end up creating the table.

vincentpoon replied:
Not sure I follow here.  My understanding is `beginCreateTable` is used to create a temporary table, then data is inserted, then `finishCreateTable` is used to rename the table so the entire operation is atomic.  So don't we need to create the table in `beginCreateTable` ?  
Phoenix doesn't support renaming of tables, so `finishCreateTable` here doesn't do anything.  Other connectors I looked at do the same (Accumulo, Cassandra, Kudu).

electrum replied:
Your understanding is correct. And you're right, I think this is the best we can do for now. We need to fix this in the connector API so that creation of the handle during planning (needed since it has to go into the query plan) is separate from the start of execution.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixMetadata.java:
```diff
@@ -0,0 +1,455 @@
+        String table = schemaTableName.getTableName();
+
+        if (!phoenixClient.getSchemaNames(JdbcIdentity.from(session)).contains(schema.orElse(null))) {
+            throw new PrestoException(NOT_FOUND, "Schema not found: " + schema);
```
Use `SchemaNotFoundException`

electrum commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixIntegrationSmokeTest.java:
```diff
@@ -0,0 +1,108 @@
+
+        assertUpdate("CREATE TABLE new_schema.test (x bigint)");
+
+        assertQueryFails("DROP SCHEMA new_schema", Pattern.quote("Error while executing statement"));
```
The quote call here should be a no-op since the pattern doesn't have any special regex characters.

Also, we should see if this can have a better error message, like `Cannot drop non-empty schema`

electrum commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixIntegrationSmokeTest.java:
```diff
@@ -0,0 +1,108 @@
+    @Test
+    public void testCreateTableWithProperties()
+    {
+        assertUpdate("CREATE TABLE test_create_table_as_if_not_exists (created_date date, a bigint, b double, c varchar(10), d varchar(10)) with(rowkeys = 'created_date row_timestamp,a,b,c', SALT_BUCKETS=10)");
```
Nit: uppercase `WITH` keyword in SQL

electrum commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixIntegrationSmokeTest.java:
```diff
@@ -0,0 +1,108 @@
+    @Test
+    public void testMultipleSomeColumnsRangesPredicate()
+    {
+        assertQuery("SELECT orderkey, shippriority, clerk, totalprice, custkey  FROM ORDERS WHERE orderkey BETWEEN 10 AND 50 or orderkey BETWEEN 100 AND 150");
```
SQL formatting nits:

* uppercase the `OR` keyword`
* lowercase the table name `ORDERS`
* remove extra space before `FROM`

electrum commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestPhoenixSqlTypeMapping.java:
```diff
@@ -0,0 +1,419 @@
+import static java.util.stream.Collectors.toList;
+
+@Test
+public class TestPhoenixSqlTypeMapping
```
We should try to de-duplicate these tests with `TestPostgreSqlTypeMapping`, but can figure that out later, as a future follow up.

electrum commented:
We have been seeing spurious test failures for `presto-tests` and `PRODUCT_TESTS_` lately. These are not related to your change.

electrum commented on presto-phoenix/src/main/java/io/prestosql/plugin/phoenix/PhoenixClient.java:
```diff
@@ -0,0 +1,450 @@
+
+    private static ResultSet getResultSet(InputSplit split, QueryPlan queryPlan)
+    {
+        List<Scan> scans = ((PhoenixInputSplit) split).getScans();
```
Make the method take `PhoenixInputSplit` rather than casting

electrum commented:
The `ConcurrentModificationException` failure for the Phoenix tests is legitimate. Here is the problem:

* `PhoenixSplit` is deserialized on a worker.
  * It contains `PhoenixInputSplit` via `WrappedPhoenixInputSplit`.
* `PhoenixInputSplit` contains a mutable `Scan` object.
  * This object is modified via the `PhoenixClient.getResultSet()` which calls `Scan.setAttribute()`.
* Before processing the table scan using the split, the engine calls `ConnectorSplit.getInfo()`, and holds onto it for later usage in logging/stats.
  * The implementation of `getInfo()` in `JdbcSplit` returns `this`, meaning the split object.
* At some later point, in a different thread, the engine then JSON serializes the returned info object. In other words, it serializes the `PhoenixSplit` object.
  * The serialization code in `WrappedPhoenixInputSplit` calls the Hadoop `Writable` serialization code in `PhoenixInputSplit`.
  * This code ends up hitting `ConcurrentModificationException`, since the `attributes` was modified concurrently.

I think the easiest fix is to have `PhoenixClient.getResultSet()` make a copy of `Scan` using the copy constructor before modifying it.

vincentpoon commented on presto-phoenix/src/test/java/io/prestosql/plugin/phoenix/TestingPhoenixServer.java:
```diff
@@ -0,0 +1,141 @@
+import static org.apache.hadoop.hbase.HConstants.MASTER_INFO_PORT;
+import static org.apache.hadoop.hbase.HConstants.REGIONSERVER_INFO_PORT;
+
+public final class TestingPhoenixServer
```
Sounds great, will do this in a followup PR

vincentpoon commented:
Thanks for the review @electrum, I've pushed a new revision that addresses the comments, and the tests are passing now.  Really appreciate all your help, I've learned a lot through the whole process.

electrum commented:
Merged, thanks!

martint commented:
Woohoo! Thanks @vincentpoon for your contribution!

findepi commented:
🎉  🚀

electrum commented:
@vincentpoon Thanks for all your hard work on this. It was a pleasure to work with you and I look forward to seeing future contributions.

vincentpoon commented:
Thank you @electrum, and thanks again @findepi and @kokosing for prior reviews.  The quality of the codebase speaks to the level of attention to detail in the code review process here.  As a result, it's been edifying and a pleasure to work on this, and I eagerly look forward to more contributions.

FYI @lhofhansl @JamesRTaylor @aertoria

vincentpoon commented:
Coincidentally, I am giving a talk at NoSql day tomorrow on this connector.  Cool that I'll be able to announce that it's been merged!
https://dataworkssummit.com/nosql-day-2019/session/integrating-apache-phoenix-with-distributed-query-engines/

electrum commented:
@vincentpoon Very cool! You can mention it will be available in the `312` release at https://prestosql.io/

lhofhansl commented:
Yeah! Thanks all!
